### PR TITLE
[Move] Implement Shed Tail

### PR DESCRIPTION
**Pull request by innerthunder**
## What are the changes the user will see?
This implements the move [Shed Tail](https://bulbapedia.bulbagarden.net/wiki/Shed_Tail_(move))

## Why am I making these changes?
Move implementations are good

## What are the changes from a developer perspective?
- `enums/switch-type`: Added a `SwitchType` enum to represent modes for switching. The current supported values are
  - `SWITCH`: Default behavior; the user can select the Pokemon to switch in.
  - `BATON_PASS`: Transferrable effects, including stat stages and tags flagged with `isBatonPassable`, are transferred from the returned Pokemon to the switched in Pokemon.
  - `SHED_TAIL`: (new) The returned Pokemon's Substitute is transferred to the switched in Pokemon.
- `data/move`:
  - Implemented Shed Tail with a combination of `AddSubstituteAttr` and `ForceSwitchOutAttr`.
  - In `ForceSwitchOutAttr`: Replaced the `batonPass` flag with a `switchType` field to add support for Shed Tail's switch type.
  - Rewrote `ForceSwitchOutAttr#apply` as a synchronous function to prevent future desync issues (i.e. it now returns a `boolean` instead of a `Promise<boolean>`)
- `phases/switch-phase` and `phases/switch-summon-phase`: Replaced `batonPass` flag with a `switchType` field to support multiple switch modes, including the new Shed Tail mode. Functions related to switching in other files have also been changed in the same vein.

### Screenshots/Videos

https://github.com/user-attachments/assets/d42e8a7d-ee42-4ed7-ae68-fb38ee15a207

## How to test the changes?
`npm run test shed_tail`

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] If I have text, did I make it translatable and add a key in the English locale file(s)?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/data/move.ts [hunk 1]
```diff
@@ -37,6 +37,7 @@ import { SwitchSummonPhase } from "#app/phases/switch-summon-phase";
 import { SpeciesFormChangeRevertWeatherFormTrigger } from "./pokemon-forms";
 import { GameMode } from "#app/game-mode";
 import { applyChallenges, ChallengeType } from "./challenge";
+import { SwitchType } from "#app/enums/switch-type";
```
Comment (DayKev) [inline src/data/move.ts:40]:
```suggestion
import { SwitchType } from "#enums/switch-type";
```

File: src/field/pokemon.ts [hunk 2]
```diff
@@ -61,6 +61,7 @@ import { Challenges } from "#enums/challenges";
 import { PokemonAnimType } from "#app/enums/pokemon-anim-type";
 import { PLAYER_PARTY_MAX_SIZE } from "#app/constants";
 import { MysteryEncounterPokemonData } from "#app/data/mystery-encounters/mystery-encounter-pokemon-data";
+import { SwitchType } from "#app/enums/switch-type";
```
Comment (DayKev) [inline src/field/pokemon.ts:64]:
```suggestion
import { SwitchType } from "#enums/switch-type";
```

File: src/phases/check-switch-phase.ts [hunk 3]
```diff
@@ -8,6 +8,7 @@ import { BattlePhase } from "./battle-phase";
 import { PostSummonPhase } from "./post-summon-phase";
 import { SummonMissingPhase } from "./summon-missing-phase";
 import { SwitchPhase } from "./switch-phase";
+import { SwitchType } from "#app/enums/switch-type";
```
Comment (DayKev) [inline src/phases/check-switch-phase.ts:11]:
```suggestion
import { SwitchType } from "#enums/switch-type";
```

File: src/phases/faint-phase.ts [hunk 4]
```diff
@@ -18,6 +18,7 @@ import { GameOverPhase } from "./game-over-phase";
 import { SwitchPhase } from "./switch-phase";
 import { VictoryPhase } from "./victory-phase";
 import { SpeciesFormChangeActiveTrigger } from "#app/data/pokemon-forms";
+import { SwitchType } from "#app/enums/switch-type";
```
Comment (DayKev) [inline src/phases/faint-phase.ts:21]:
```suggestion
import { SwitchType } from "#enums/switch-type";
```

File: src/phases/mystery-encounter-phases.ts [hunk 5]
```diff
@@ -24,6 +24,7 @@ import { NewBattlePhase } from "#app/phases/new-battle-phase";
 import { GameOverPhase } from "#app/phases/game-over-phase";
 import { SwitchPhase } from "#app/phases/switch-phase";
 import { SeenEncounterData } from "#app/data/mystery-encounters/mystery-encounter-save-data";
+import { SwitchType } from "#app/enums/switch-type";
```
Comment (DayKev) [inline src/phases/mystery-encounter-phases.ts:27]:
```suggestion
import { SwitchType } from "#enums/switch-type";
```

File: src/phases/return-phase.ts [hunk 6]
```diff
@@ -1,10 +1,11 @@
 import BattleScene from "#app/battle-scene";
 import { SpeciesFormChangeActiveTrigger } from "#app/data/pokemon-forms";
+import { SwitchType } from "#app/enums/switch-type";
```
Comment (DayKev) [inline src/phases/return-phase.ts:3]:
```suggestion
import { SwitchType } from "#enums/switch-type";
```

File: src/phases/switch-phase.ts [hunk 7]
```diff
@@ -1,4 +1,5 @@
 import BattleScene from "#app/battle-scene";
+import { SwitchType } from "#app/enums/switch-type";
```
Comment (DayKev) [inline src/phases/switch-phase.ts:2]:
```suggestion
import { SwitchType } from "#enums/switch-type";
```

File: src/phases/switch-summon-phase.ts [hunk 8]
```diff
@@ -12,29 +12,30 @@ import i18next from "i18next";
 import { PostSummonPhase } from "./post-summon-phase";
 import { SummonPhase } from "./summon-phase";
 import { SubstituteTag } from "#app/data/battler-tags";
+import { SwitchType } from "#app/enums/switch-type";
```
Comment (DayKev) [inline src/phases/switch-summon-phase.ts:15]:
```suggestion
import { SwitchType } from "#enums/switch-type";
```

File: src/phases/turn-start-phase.ts [hunk 9]
```diff
@@ -19,6 +19,7 @@ import { TurnEndPhase } from "./turn-end-phase";
 import { WeatherEffectPhase } from "./weather-effect-phase";
 import { BattlerIndex } from "#app/battle";
 import { TrickRoomTag } from "#app/data/arena-tag";
+import { SwitchType } from "#app/enums/switch-type";
```
Comment (DayKev) [inline src/phases/turn-start-phase.ts:22]:
```suggestion
import { SwitchType } from "#enums/switch-type";
```

File: src/phases/turn-start-phase.ts [hunk 10]
```diff
@@ -179,7 +180,8 @@ export class TurnStartPhase extends FieldPhase {
         this.scene.unshiftPhase(new AttemptCapturePhase(this.scene, turnCommand.targets![0] % 2, turnCommand.cursor!));//TODO: is the bang correct here?
         break;
       case Command.POKEMON:
-        this.scene.unshiftPhase(new SwitchSummonPhase(this.scene, pokemon.getFieldIndex(), turnCommand.cursor!, true, turnCommand.args![0] as boolean, pokemon.isPlayer()));//TODO: is the bang correct here?
+        const switchType = turnCommand.args![0] ? SwitchType.BATON_PASS : SwitchType.SWITCH; // TODO: is the bang correct here?
```
Comment (DayKev) [inline src/phases/turn-start-phase.ts:183]:
```suggestion
        const switchType = turnCommand.args?.[0] ? SwitchType.BATON_PASS : SwitchType.SWITCH;
```

File: src/test/abilities/zen_mode.test.ts [hunk 11]
```diff
@@ -18,6 +18,7 @@ import GameManager from "#test/utils/gameManager";
 import Phaser from "phaser";
 import { afterEach, beforeAll, beforeEach, describe, expect, test } from "vitest";
 import { Status, StatusEffect } from "#app/data/status-effect";
+import { SwitchType } from "#app/enums/switch-type";
```
Comment (DayKev) [inline src/test/abilities/zen_mode.test.ts:21]:
```suggestion
import { SwitchType } from "#enums/switch-type";
```

**Review by DayKev — COMMENTED** [review]
Thanks vscode.

File: src/data/move.ts [hunk 12]
```diff
@@ -1521,6 +1526,66 @@ export class AddSubstituteAttr extends MoveEffectAttr {
   }
 }
 
+/**
+ * Attribute implementing the effects of {@link https://bulbapedia.bulbagarden.net/wiki/Shed_Tail_(move) | Shed Tail}.
+ * This is essentially a wrapper for two internal attributes: {@linkcode AddSubstituteAttr} which handles creating
+ * a Substitute for the user, and {@linkcode ForceSwitchOutAttr} which forces the user to switch out and transfers the
+ * Substitute to the switched in Pokemon.
+ */
+export class ShedTailAttr extends MoveEffectAttr {
```
Comment (DayKev) [inline src/data/move.ts:1535]: Why did you need to create a mixin Attr? The move should fail if either Attr's condition doesn't pass, which is the current behavior of moves. It seems like you should just be able to put `.attr(AddSubstituteAttr, 0.5).attr(ForceSwitchOutAttr, true, SwitchType.SHED_TAIL),` on the move definition.

Comment (DayKev) [reply src/data/move.ts:1535]: Quick testing indicates that it should work just fine this way, as well.

Comment (innerthunder) [reply src/data/move.ts:1535]: I think I made it before I changed `ForceSwitchOutAttr`'s apply function away from a Promise, and this just kinda stuck. If it would work either way, I'll go ahead and remove it

**Review by DayKev — APPROVED** [review]

**Review by Adri1 — APPROVED** [review]

**Review by Adri1 — DISMISSED** [review]

**Review by DayKev — APPROVED** [review]

File: src/data/move.ts [hunk 13]
```diff
@@ -1493,8 +1499,7 @@ export class AddSubstituteAttr extends MoveEffectAttr {
       return false;
     }
 
-    const hpCost = Math.floor(user.getMaxHp() / 4);
-    user.damageAndUpdate(hpCost, HitResult.OTHER, false, true, true);
+    user.damageAndUpdate(Math.floor(user.getMaxHp() * this.hpCost), HitResult.OTHER, false, true, true);
```
Comment (torranx) [inline src/data/move.ts:1502]: shouldn't this be using `Math.max()`?

File: src/data/move.ts [hunk 14]
```diff
@@ -1507,7 +1512,7 @@ export class AddSubstituteAttr extends MoveEffectAttr {
   }
 
   getCondition(): MoveConditionFunc {
-    return (user, target, move) => !user.getTag(SubstituteTag) && user.hp > Math.floor(user.getMaxHp() / 4) && user.getMaxHp() > 1;
+    return (user, target, move) => !user.getTag(SubstituteTag) && user.hp > Math.floor(user.getMaxHp() * this.hpCost) && user.getMaxHp() > 1;
```
Comment (torranx) [inline src/data/move.ts:1515]: same here

File: src/data/move.ts [hunk 15]
```diff
@@ -5304,7 +5306,7 @@ export class ChillyReceptionAttr extends ForceSwitchOutAttr {
 
   // using inherited constructor
```
Comment (torranx) [inline src/data/move.ts:5307]: can we just remove this?

File: src/phases/switch-phase.ts [hunk 16]
```diff
@@ -10,21 +11,24 @@ import { SwitchSummonPhase } from "./switch-summon-phase";
...
 export class SwitchPhase extends BattlePhase {
   protected fieldIndex: integer;
+  private switchType: SwitchType;
   private isModal: boolean;
   private doReturn: boolean;
 
   /**
      * Creates a new SwitchPhase
      * @param scene {@linkcode BattleScene} Current battle scene
+     * @param switchType The type of switch logic this phase implements
```
Comment (torranx) [inline src/phases/switch-phase.ts:21]:
```suggestion
     * @param switchType {@linkcode SwitchType} The type of switch logic this phase implements
```

File: src/data/move.ts (hunk 13 above)
Comment (DayKev) [reply src/data/move.ts:1502]: > The user takes 1/4 of its maximum HP, rounded down
- https://www.smogon.com/dex/sv/moves/substitute/

Comment (torranx) [reply src/data/move.ts:1502]: I mean `Math.max(1, Math.floor())`, in the case where pokemon is Shedinja (1 hp)

Comment (innerthunder) [reply src/data/move.ts:1502]: The move fails when `user.getMaxHp() === 1` (see `getCondition`), so that case never gets to this point

Comment (torranx) [reply src/data/move.ts:1502]: I think you meant `user.getMaxHp() > 1` - consider this comment resolved 👍

Comment (torranx) [reply src/data/move.ts:1502]: wait isn't this the [condition](https://github.com/pagefaultgames/pokerogue/pull/4382/files#diff-62e9cf38a6ba2a2b69a102788cdfef2a33cdc173f79124df114636685a51e762R1515)? 
`return (user, target, move) => !user.getTag(SubstituteTag) && user.hp > Math.floor(user.getMaxHp() * this.hpCost) && user.getMaxHp() > 1;`

Comment (innerthunder) [reply src/data/move.ts:1502]: Yeah, so it fails/returns false when `user.getMaxHp() <= 1`

**Review by torranx — DISMISSED** [review]
looks good

File: src/data/move.ts (hunk 14 above)
Comment (innerthunder) [reply src/data/move.ts:1515]: sorted out in the above comment thread

**Review by Chapybara-jp — APPROVED** [review]

**Review by flx-sta — APPROVED** [review]

**Review by Chapybara-jp — APPROVED** [review]

**Review by DayKev — APPROVED** [review]

**Review by torranx — APPROVED** [review]

**Review by Adri1 — APPROVED** [review]
